### ISSUE TITLE
Ingress capabilities detection

### ISSUE THREAD

**Issue by ludovic-pourrat** [Kong/charts#616]
Hi,

Since we upgraded our Kubernetes cluster we identified that the `kong.ingressVersion` helper function is not applying the right capabilities detection switch.

We are automatically detecting capabilities on our cluster and passing those to the Helm templating via `kubectl api-versions`

Current implementation defines

```
{{- define "kong.ingressVersion" -}}
{{- if (.Capabilities.APIVersions.Has "networking.k8s.io/v1/Ingress") -}}
networking.k8s.io/v1
{{- else if (.Capabilities.APIVersions.Has "networking.k8s.io/v1beta1/Ingress") -}}
networking.k8s.io/v1beta1
{{- else -}}
extensions/v1beta1
{{- end -}}
{{- end -}}
```

Where it should be 

```
{{- define "kong.ingressVersion" -}}
{{- if (.Capabilities.APIVersions.Has "networking.k8s.io/v1") -}}
networking.k8s.io/v1
{{- else if (.Capabilities.APIVersions.Has "networking.k8s.io/v1beta1") -}}
networking.k8s.io/v1beta1
{{- else -}}
extensions/v1beta1
{{- end -}}
{{- end -}}
```

Are our assumptions correct ?

If so I will raise a PR for this.

Many thanks.

**Comment by pmalek** [comment]
Hi @ludovic-pourrat 👋 

It seems you're right. Given this handy template:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: apiversions-{{ .Release.Name }}
  namespace: apiversions
data:
  apiVersions.txt: |
    {{- range .Capabilities.APIVersions }}
    - {{ . }}
    {{- end -}}
```

I can confirm I'm getting the api versions as you've indicated:

```yaml
---
# Source: kong/templates/cm.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: apiversions-kong-test1
  namespace: apiversions
data:
  apiVersions.txt: |
    - v1
    - admissionregistration.k8s.io/v1
    - admissionregistration.k8s.io/v1beta1
    - internal.apiserver.k8s.io/v1alpha1
    - apps/v1
    - apps/v1beta1
    - apps/v1beta2
    - authentication.k8s.io/v1
    - authentication.k8s.io/v1beta1
    - authorization.k8s.io/v1
    - authorization.k8s.io/v1beta1
    - autoscaling/v1
    - autoscaling/v2
    - autoscaling/v2beta1
    - autoscaling/v2beta2
    - batch/v1
    - batch/v1beta1
    - certificates.k8s.io/v1
    - certificates.k8s.io/v1beta1
    - coordination.k8s.io/v1beta1
    - coordination.k8s.io/v1
    - discovery.k8s.io/v1
    - discovery.k8s.io/v1beta1
    - events.k8s.io/v1
    - events.k8s.io/v1beta1
    - extensions/v1beta1
    - flowcontrol.apiserver.k8s.io/v1alpha1
    - flowcontrol.apiserver.k8s.io/v1beta1
    - flowcontrol.apiserver.k8s.io/v1beta2
    - networking.k8s.io/v1
    - networking.k8s.io/v1beta1
    - node.k8s.io/v1
    - node.k8s.io/v1alpha1
    - node.k8s.io/v1beta1
    - policy/v1
    - policy/v1beta1
    - rbac.authorization.k8s.io/v1
    - rbac.authorization.k8s.io/v1beta1
    - rbac.authorization.k8s.io/v1alpha1
    - scheduling.k8s.io/v1alpha1
    - scheduling.k8s.io/v1beta1
    - scheduling.k8s.io/v1
    - storage.k8s.io/v1beta1
    - storage.k8s.io/v1
    - storage.k8s.io/v1alpha1
    - apiextensions.k8s.io/v1beta1
    - apiextensions.k8s.io/v1
```

Feel free to submit a fix for that.

**Comment by mflendrich** [comment]
@ludovic-pourrat thanks for reporting :bow:.
~Would you be interested in submitting a PR?~

**Comment by mflendrich** [comment]
@ludovic-pourrat 
After discussion with the team: before jumping to a fix, could you please let us know what exact issue has this caused for you?

**Comment by pmalek** [comment]
BTW it seems that we can't query particular objects APIs via `.Capabilities.APIVersions`

e.g. having:

```
$ k get --raw /apis/networking.k8s.io/v1 | jq '.resources[] | select(.kind == "Ingress")'
{
  "name": "ingresses",
  "singularName": "",
  "namespaced": true,
  "kind": "Ingress",
  "verbs": [
    "create",
    "delete",
    "deletecollection",
    "get",
    "list",
    "patch",
    "update",
    "watch"
  ],
  "shortNames": [
    "ing"
  ],
  "storageVersionHash": "39NQlfNR+bo="
}
{
  "name": "ingresses/status",
  "singularName": "",
  "namespaced": true,
  "kind": "Ingress",
  "verbs": [
    "get",
    "patch",
    "update"
  ]
}
```

`{{ (.Capabilities.APIVersions.Has "networking.k8s.io/v1/Ingress" ) }}` returns `false`

**Comment by ludovic-pourrat** [comment]
> @ludovic-pourrat 
> 
> After discussion with the team: before jumping to a fix, could you please let us know what exact issue has this caused for you?

We cannot deploy anymore Kong because the default to the release of betav1 is not anymore available in our migrated cluster.

**Comment by rainest** [comment]
Querying for a specific resource works fine. [Helm's documentation](https://helm.sh/docs/chart_template_guide/function_list/#capabilitiesapiversionshas) indicates that `Has` supports either. Not quite sure how to explain the range output; I haven't gone into the actual implementation.

We use the specific resources because resources in a group don't iterate versions in lockstep, and there are Kubernetes versions where there are are some `[networking.k8s.io](http://networking.k8s.io/)` resources at `v1` while Ingress is still only available at `v1beta1`:

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#networkpolicy-v1-networking-k8s-io

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ingress-v1beta1-networking-k8s-io
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ingress-v1beta1-extensions

@ludovic-pourrat between the lack of guarantee of specific resource for an available API version and the template behavior observed below, you'll need to rework your script to parse `kubectl api-resources` output and build `--api-versions` arguments from that. Changing the templates to use the Group+Kind only would (at least on recent versions of Helm) render networking v1 Ingresses regardless of whether they're actually available on the cluster.

A few tests with [an expanded version of the ConfigMap template from earlier](https://github.com/Kong/charts/files/8951634/configmap.yaml.txt) shows a few interesting things. An actual install with Helm interrogating the (1.23) cluster does correctly detect the availability of v1 Ingress resources:

```
11:10:49-0700 letov $ helm install --dry-run example ./  | grep -A5 apiversions-test
  name: apiversions-test
  namespace: apiversions
data:
  apiVersions.txt: |
    networking.k8s.io/v1/Ingress
    networking.k8s.io/v1
11:10:55-0700 letov $ kubectl get --raw /apis/networking.k8s.io/v1 | jq '.resources[] | select(.kind == "Ingress") | .name'
"ingresses"
"ingresses/status"
```

Running `helm template` shows some odd behavior where it apparently has a built-in set of versions it assumes exist. Checks for Ingress detect nothing and fall back to extensions v1beta1, but the check for just the API version chooses the highest available, networking v1:

```
1:15:52-0700 letov $ helm template ./ | grep -A5 apiversions-test
  name: apiversions-test
  namespace: apiversions
data:
  apiVersions.txt: |
    extensions/v1beta1/Ingress
    networking.k8s.io/v1
```
Providing a higher Ingress version using `--api-versions` works as expected:

```
11:16:03-0700 letov $ helm template ./ --api-versions networking.k8s.io/v1beta1/Ingress | grep -A5 apiversions-test
  name: apiversions-test
  namespace: apiversions
data:
  apiVersions.txt: |
    networking.k8s.io/v1beta1/Ingress
    networking.k8s.io/v1

11:16:10-0700 letov $ helm template ./ --api-versions networking.k8s.io/v1/Ingress | grep -A5 apiversions-test
  name: apiversions-test
  namespace: apiversions
data:
  apiVersions.txt: |
    networking.k8s.io/v1/Ingress
    networking.k8s.io/v1
```
Non-standard API groups have no built-in API version knowledge, and notably providing a resource with a given Group+Version does not return true for a `Has` query for the Group+Version only, while it does return true for the resource itself:

```
11:16:29-0700 letov $ helm template ./ --api-versions fakegroup.example.com/v1/FakeResource | grep -A7 apiversions-test
  name: apiversions-test
  namespace: apiversions
data:
  apiVersions.txt: |
    extensions/v1beta1/Ingress
    networking.k8s.io/v1
    fakegroup.example.com/v1/FakeResource
---
11:18:00-0700 letov $ helm template ./ --api-versions fakegroup.example.com/v1 | grep -A7 apiversions-test
  name: apiversions-test
  namespace: apiversions
data:
  apiVersions.txt: |
    extensions/v1beta1/Ingress
    networking.k8s.io/v1

    fakegroup.example.com/v1
```

**Comment by ludovic-pourrat** [comment]
Am aside note we are running Kubernetes 1.23 where beta/v1 is dropped.

**Comment by ludovic-pourrat** [comment]
Hi @rainest 

Thanks for your feedback.

1. No issue on our side to change our rendering scripts to pass a custom `api-versions`
2. I would rather think that it is better to rely on builtin api versions instead of creating custom ones
3. Thw 1.19 is out of support so waht is the point of no switching to builtin api versions ?

Many thanks.

**Comment by rainest** [comment]
Not sure I follow your question about the built-in versus custom API versions. Are you referring to the Helm behavior or to the `fakegroup.example.com/v1` version?

I included the `fakegroup.example.com/v1` to demonstrate the difference between Helm's behavior for standard Kubernetes groups (Helm appears to have built-in info about assumed available groups when running `helm template` without any `--api-versions` arguments) and custom ones. The custom ones aren't actually relevant here, though we do use them elsewhere, e.g. to detect whether we can create Prometheus resources.

We've kept support for 1.19 resources in part out of a lack of compelling reason to remove them and in part because we do observe a number of still active 1.19 deployments in the wild. Absent something else that will break compatibility, we tend to try and keep compatibility with with older versions if possible, and it was easy to do so here because we'd already written the template logic when 1.19 was still a supported version.

The Helm built-in knowledge is also incomplete: it's aware of the group and version (though it's not clear what Kubernetes version it bases these on), but not individual kinds within it, and Kubernetes can introduce new beta version kinds into an existing group that has v1 APIs already. The built-in versions furthermore aren't guaranteed to match the actual destination cluster's, since they're determined entirely by the Helm version. IMO the ideal solution here is for the major users of template (various CD solutions) to find a way to essentially mimic the data normally available to Helm during `install` runs and populate this automatically--individual chart authors shouldn't need to design for two modes of operation that function differently. Helm has more or less indicated that they'd prefer template users [switch to an SDK-based implementation](https://github.com/helm/helm/pull/10108) rather than adding additional `template` CLI functionality.

**Comment by ludovic-pourrat** [comment]
@rainest,

Sorry I was not clear enough in my explanations for the built-in options in Helm, and I was referring the one when you pass the `api-versions` to the templating.

Our deployment process execute a `kubectl api-versions` and pass it to Helm in order to template the files as we are using Argo  for the deployment.

What is not clear to me is how this is supposed to work in a Kubernetes release greater than 1.19 ? (if not using the api-versions).

**Comment by rainest** [comment]
This sort of issue only appears with `helm template`, not with `helm install`. The latter is what we've historically targeted, because that is the standard means of installing a Helm chart. My understanding of the situation (to be taken with a grain of salt--I haven't discussed this with the stakeholders) is roughly:

- Helm was initially designed with `install` as the install tool and `template` as mostly a debugging utility.
- Various CD tools (Argo ageneralnd such) started adding Helm install methods because of the wide availability of charts for applications. Because they wanted to manage applying resources themselves, they used `helm template` instead of `helm install`. 
- `template` isn't really well-suited for this purpose, because it does a lot of things you probably wouldn't want for it, namely rendering test and lifecycle hook templates by default, not applying the namespace, and breaking anything that requires talking to the target Kubernetes cluster.
- The Helm project does not want to add all functionality needed to make `template` work well for the CD case to the command, because they consider it a tool designed for humans, whereas integrations should instead use the Helm Go API (which should essentially allow them a path to doing the useful things `install` does without losing control over resource installation).

The two parties have not resolved this, so we (chart authors) end up stuck in the awkward situation of needing to either offer annoying workarounds or add additional toggles to our values.yaml to replicate functionality normally provided by Helm.

**Comment by ludovic-pourrat** [comment]
@rainest 

Many thanks for your feedback ! We will amend our pipeline to workaround this by passing a custom `api-versions` to include the ingress resource.

We will revisit this approach once a cleaner solution will be available, or once the support for 1.19 release of Kubernetes will be dropped.
